File: longest-palindromic-substring/cpp/substring.cpp
#include <iostream>
#include <string>

using namespace std;

class Solution {
public:
    // Yordamchi funksiya: Markazdan kengaytirib eng uzun palindrom substringni topish
    string expandAroundCenter(const string& s, int left, int right) {
        // Chap va o'ng tomondan kengaytirish
        while (left >= 0 && right < s.length() && s[left] == s[right]) {
            left--;
            right++;
        }
        // Oxirgi to'g'ri palindrom substringni qaytarish
        return s.substr(left + 1, right - left - 1);
    }

    string longestPalindrome(string s) {
        // Agar matn uzunligi 1 yoki kam bo'lsa, o'zi palindrom
        if (s.length() < 2) {
            return s;
        }

        string longestPalindrome = "";
        for (int i = 0; i < s.length(); i++) {
            // 1. Belgini markaz qilib kengaytirish
            string oddPalindrome = expandAroundCenter(s, i, i);
            // 2. Harf juftligini markaz qilib kengaytirish
            string evenPalindrome = expandAroundCenter(s, i, i + 1);

            // Eng uzun palindromni yangilash
            if (oddPalindrome.length() > longestPalindrome.length()) {
                longestPalindrome = oddPalindrome;
            }
            if (evenPalindrome.length() > longestPalindrome.length()) {
                longestPalindrome = evenPalindrome;
            }
        }

        return longestPalindrome;
    }
};

// Misollarni tekshirish
int main() {
    Solution solution;
    cout << solution.longestPalindrome("babad") << endl;  // Output: "bab" yoki "aba"
    cout << solution.longestPalindrome("cbbd") << endl;   // Output: "bb"
    return 0;
}

File: longest-palindromic-substring/python2/substring.py
class Solution:
    def longestPalindrome(self, s):
        def expandAroundCenter(left, right):
            while left >= 0 and right < len(s) and s[left] == s[right]:
                left -= 1
                right += 1
            return s[left + 1:right]
        
        if not s:
            return ""

        longest_palindrome = ""
        for i in range(len(s)):
            # Check for odd-length palindromes
            odd_palindrome = expandAroundCenter(i, i)
            if len(odd_palindrome) > len(longest_palindrome):
                longest_palindrome = odd_palindrome
            
            # Check for even-length palindromes
            even_palindrome = expandAroundCenter(i, i + 1)
            if len(even_palindrome) > len(longest_palindrome):
                longest_palindrome = even_palindrome
        
        return longest_palindrome